### ISSUE TITLE
F821 false positive when assigning to a dict

### ISSUE THREAD

**Issue by jiwidi** [charliermarsh/ruff#580]
Hi! Love what you are building with ruff. 

I'm trying it out now and found a weird behavior that i dont know if its a bug or something im missing.

```python
def api_event_json(api_key, model_name, user_id=None, item_id=None):
    dict = {
        "headers": {
            "x-api-key": api_key,
        },
        "queryStringParameters": {"limit": "5"},
        "pathParameters": {"model": model_name},
    }

    if user_id is not None:
        dict["queryStringParameters"]["user_id"] = user_id
    if item_id is not None:
        dict["queryStringParameters"]["item_id"] = item_id

    return dict
```

Why would this code throw F821? 

```
F821 Undefined name `queryStringParameters`
F821 Undefined name `queryStringParameters`
```

For the lines where we access the dict.

I can walk around this error by rearranging the code a bit:
```python
def api_event_json(api_key, model_name, user_id=None, item_id=None):

    query_params = {"limit": 5}
    if user_id is not None:
        query_params["user_id"] = user_id
    if item_id is not None:
        query_params["item_id"] = item_id

    dict = {
        "headers": {
            "x-api-key": api_key,
        },
        "queryStringParameters": query_params,
        "pathParameters": {"model": model_name},
    }

    return dict
```

And is alright. Wondering if its a bug.

**Comment by charliermarsh** [comment]
Definitely a bug! It’s treating it like a type annotation with a forward reference (eg, x: dict[“foo”] = {}, which would be referencing a variable named “foo”). But it should only be enforcing that behavior for annotations. Will fix today.

**Comment by charliermarsh** [comment]
Another way to fix this is to use a name other than `dict`. Part of the problem is that the code is shadowing the built-in `dict`.

**Comment by charliermarsh** [comment]
This'll go out in the next release, later today!
